readme.md
# Custom Document Well for Visual Studio Code

## About
A replication a subset of Visual Studio's (discontinued)
[Custom Document Well](https://marketplace.visualstudio.com/items?itemName=VisualStudioPlatformTeam.CustomDocumentWell)
plugin.

The standard VS Code extension API does not support this style of modification,
this is instead done by direct DOM manipulation. This causes **an additional
installation step to be required.** See [Installation](#installation).

**Note:** Only the side bar being on the left is supported at this time.

Supported functionality:
* Vertical tabs (left side only at present).
* Sort and colorize by project.
* Sort by file type.

## Installation
1. Install the VS Code extension.
2. `View` -> `Command Palette`, "Install Custom Document Well" `{enter}`
3. Restart VS Code. **It may need to be restarted as an administrator!** See notes for [Custom CSS and JS Loader](https://marketplace.visualstudio.com/items?itemName=be5invis.vscode-custom-css)

Alternatively, this can be more easily run non-persistant on a single instance:
1. Open developer tools in VS Code: `Help` -> `Toggle Developer Tools`
2. Paste the contents of [customdocumentwell.js](https://github.com/jlennox/VSCode-CustomDocumentWell/blob/master/out/customdocumentwell.js) into the Console. Press enter.

## Development
1. Clone
2. Enter the project directory in `cmd.exe` or equivalent.
3. Enter tsc's watch with `node "C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsc.js" --watch` or equivalent.
4. Press `F5` to start debugging in VS Code.
5. Run "Install Custom Document Well" command if not yet run. Only needs to be done once.

Alternatively, it can be loaded by pasting the .js into the developer console,
which can be easier for rapid development and when introducing potentionally
view breaking bugs.

## Remaining work/rough corners.

The vertical tabs can not be repositioned or resized. It's always on the left
and always 300px wide.

This was not done as a PR to VS Code because the changes and features are
personal and unlikely to be intune with the intended VS Code product direction.

## Screenshots

![Screenshot](/screenshots/sample-screenshot-a.png?raw=true)

![Screenshot](/screenshots/sample-screenshot-b.png?raw=true)

CHANGELOG.md
# Change Log


## ????
* **Feature:** Ability to have an expression that causes tabs to be underlined when matched. Defaults to matching files in node_module directories.

## 1.2.1
* **Feature:** Fix 1.53.2 support. Still requires manually editing the workbench.js to permit js injection, sadly.

## 1.2.0
* **Feature:** Pinned tabs now save between sessions.
* **Bug:** Fix that some tabs weren't sorting/grouping by path correctly. It was possible for the path separator to be mixed.

## 1.1.0
* **Feature:** Add tab "pinning."
* **Feature:** Allow project detection setting.
* **Bug:** Fix "peek" preview not displaying correctly.

out/extension.js
"use strict";
// The MIT License(MIT)
//
// Copyright(c) 2016 Roberto Huertas
// Copyright(c) 2016 Belleve Invis
// https://github.com/be5invis/vscode-custom-css
// https://github.com/be5invis/vscode-custom-css/blob/master/LICENSE.txt
//
// Modifications:
// The MIT License(MIT)
// Copyright(c) 2020 Joseph Lennox
Object.defineProperty(exports, "__esModule", { value: true });
const vscode = require("vscode");
const path = require("path");
const events = require("events");
const fs = require("fs");
const messages_1 = require("./messages");
function activate(context) {
    var _a, _b;
    process.on("uncaughtException", (error) => {
        if (/ENOENT|EACCES|EPERM/.test(error.code || "")) {
            vscode.window.showInformationMessage(messages_1.messages.admin);
            return;
        }
    });
    const eventEmitter = new events.EventEmitter();
    const appPath = (_b = (_a = require) === null || _a === void 0 ? void 0 : _a.main) === null || _b === void 0 ? void 0 : _b.filename;
    if (!appPath) {
        console.error("Custom Document Well: Unable to locate application path.");
        return;
    }
    const appDir = path.dirname(appPath);
    const base = path.join(appDir, "vs", "code");
    const htmlFile = path.join(base, "electron-browser", "workbench", "workbench.html");
    const htmlFileBack = path.join(base, "electron-browser", "workbench", "workbench.bak-cdw");
    function replaceCss() {
        mutateFile((contents) => {
            contents = removeInjectedHtml(contents);
            return injectHtml(contents);
        });
    }
    function removeInjectedHtml(contents) {
        return contents.replace(/<!-- !! VSCODE-CDW-START !! -->[\s\S]*?<!-- !! VSCODE-CDW-END !! -->/, "");
    }
    function injectHtml(contents) {
        const config = vscode.workspace.getConfiguration("custom_document_well");
        let cdwPath = path.join(__dirname, "customdocumentwell.js");
        if (path.sep === "\\") {
            cdwPath = cdwPath.replace(/\\/g, "/");
        }
        const injectHTML = `
			<span id="__hack_cdw_config" style="display: none;">${encodeURI(JSON.stringify(config))}</span>
			<script src="file:///${cdwPath}" type="application/javascript"></script>`;
        return contents.replace("</html>", `<!-- !! VSCODE-CDW-START !! -->${injectHTML}<!-- !! VSCODE-CDW-END !! --></html>`);
    }
    function mutateFile(cb) {
        try {
            const originalHtml = fs.readFileSync(htmlFile, "utf-8");
            const html = cb(originalHtml);
            if (originalHtml == html)
                return;
            fs.writeFileSync(htmlFile, html, "utf-8");
            enabledRestart();
        }
        catch (e) {
            console.log(e);
        }
    }
    function timeDiff(d1, d2) {
        return Math.abs(d2.getTime() - d1.getTime());
    }
    function hasBeenUpdated(stats1, stats2) {
        const dbak = new Date(stats1.ctime);
        const dor = new Date(stats2.ctime);
        const segs = timeDiff(dbak, dor) / 1000;
        return segs > 60;
    }
    function cleanCssInstall() {
        const c = fs
            .createReadStream(htmlFile)
            .pipe(fs.createWriteStream(htmlFileBack));
        c.on("finish", function () {
            replaceCss();
        });
    }
    function installItem(bakfile, orfile, cleanInstallFunc) {
        fs.stat(bakfile, function (errBak, statsBak) {
            //if (errBak)
            {
                // clean installation
                cleanInstallFunc();
                //  return;
            }
            // check htmlFileBack's timestamp and compare it to the htmlFile's.
            fs.stat(orfile, function (errOr, statsOr) {
                if (errOr) {
                    vscode.window.showInformationMessage(messages_1.messages.smthingwrong + errOr);
                    return;
                }
                var updated = hasBeenUpdated(statsBak, statsOr);
                if (updated) {
                    // some update has occurred. clean install
                    cleanInstallFunc();
                }
            });
        });
    }
    function emitEndUninstall() {
        eventEmitter.emit("endUninstall");
    }
    function restoredAction(isRestored, willReinstall) {
        if (isRestored >= 1) {
            if (willReinstall) {
                emitEndUninstall();
            }
            else {
                disabledRestart();
            }
        }
    }
    function restoreBak(willReinstall = false) {
        var restore = 0;
        fs.unlink(htmlFile, function (err) {
            if (err) {
                vscode.window.showInformationMessage(messages_1.messages.admin);
                return;
            }
            var c = fs
                .createReadStream(htmlFileBack)
                .pipe(fs.createWriteStream(htmlFile));
            c.on("finish", function () {
                fs.unlinkSync(htmlFileBack);
                restore++;
                restoredAction(restore, willReinstall);
            });
        });
    }
    function reloadWindow() {
        // reload vscode-window
        vscode.commands.executeCommand("workbench.action.reloadWindow");
    }
    function enabledRestart() {
        vscode.window
            .showInformationMessage(messages_1.messages.enabled, { title: messages_1.messages.restartIde })
            .then(function (msg) {
            reloadWindow();
        });
    }
    function disabledRestart() {
        vscode.window
            .showInformationMessage(messages_1.messages.disabled, { title: messages_1.messages.restartIde })
            .then(function (msg) {
            reloadWindow();
        });
    }
    // ####  main commands ######################################################
    function fInstall() {
        installItem(htmlFileBack, htmlFile, cleanCssInstall);
    }
    function fUninstall(willReinstall = false) {
        fs.stat(htmlFileBack, function (errBak, statsBak) {
            if (errBak) {
                if (willReinstall) {
                    emitEndUninstall();
                }
                return;
            }
            fs.stat(htmlFile, function (errOr, statsOr) {
                if (errOr) {
                    vscode.window.showInformationMessage(messages_1.messages.smthingwrong + errOr);
                    return;
                }
                restoreBak(willReinstall);
            });
        });
    }
    function fUpdate() {
        eventEmitter.once("endUninstall", fInstall);
        fUninstall(true);
    }
    context.subscriptions.push(vscode.commands.registerCommand("extension.installCustomDocumentWell", fInstall));
    context.subscriptions.push(vscode.commands.registerCommand("extension.uninstallCustomDocumentWell", fUninstall));
    context.subscriptions.push(vscode.commands.registerCommand("extension.updateCustomDocumentWell", fUpdate));
}
exports.activate = activate;
// this method is called when your extension is deactivated
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map

src/extension.ts
// The MIT License(MIT)
//
// Copyright(c) 2016 Roberto Huertas
// Copyright(c) 2016 Belleve Invis
// https://github.com/be5invis/vscode-custom-css
// https://github.com/be5invis/vscode-custom-css/blob/master/LICENSE.txt
//
// Modifications:
// The MIT License(MIT)
// Copyright(c) 2020 Joseph Lennox

interface Error
{
    code?: string;
    stack?: string;
}

import * as vscode from 'vscode';
import path = require("path");
import events = require("events");
import fs = require("fs");
import { messages as msg } from "./messages";

export function activate(context: vscode.ExtensionContext): void
{
    process.on("uncaughtException", (error: Error): void =>
    {
        if (/ENOENT|EACCES|EPERM/.test(error.code || ""))
        {
            vscode.window.showInformationMessage(msg.admin);
            return;
        }
    });

    const eventEmitter = new events.EventEmitter();
    const appPath = require?.main?.filename;

    if (!appPath)
    {
        console.error("Custom Document Well: Unable to locate application path.");
        return;
    }

    const appDir = path.dirname(appPath);
    const base = path.join(appDir, "vs", "code");
    const htmlFile = path.join(base, "electron-browser", "workbench", "workbench.html");
    const htmlFileBack = path.join(base, "electron-browser", "workbench", "workbench.bak-cdw");

    function replaceCss()
    {
        mutateFile((contents: string) =>
        {
            contents = removeInjectedHtml(contents);
            return injectHtml(contents);
        });
    }

    function removeInjectedHtml(contents: string): string
    {
        return contents.replace(
            /<!-- !! VSCODE-CDW-START !! -->[\s\S]*?<!-- !! VSCODE-CDW-END !! -->/,
            "");
    }

    function injectHtml(contents: string): string
    {
        const config = vscode.workspace.getConfiguration("custom_document_well");
        let cdwPath = path.join(__dirname, "customdocumentwell.js");

        if (path.sep === "\\")
        {
            cdwPath = cdwPath.replace(/\\/g, "/");
        }

        const injectHTML = `
			<span id="__hack_cdw_config" style="display: none;">${encodeURI(JSON.stringify(config))}</span>
			<script src="file:///${cdwPath}" type="application/javascript"></script>`;

        return contents.replace(
            "</html>",
            `<!-- !! VSCODE-CDW-START !! -->${injectHTML}<!-- !! VSCODE-CDW-END !! --></html>`);
    }

    function mutateFile(cb: (contents: string) => string): void
    {
        try
        {
            const originalHtml = fs.readFileSync(htmlFile, "utf-8");
            const html = cb(originalHtml);

            if (originalHtml == html) return;

            fs.writeFileSync(htmlFile, html, "utf-8");
            enabledRestart();
        }
        catch (e)
        {
            console.log(e);
        }
    }

    function timeDiff(d1: Date, d2: Date): number
    {
        return Math.abs(d2.getTime() - d1.getTime());
    }

    function hasBeenUpdated(stats1: fs.Stats, stats2: fs.Stats): boolean
    {
        const dbak = new Date(stats1.ctime);
        const dor = new Date(stats2.ctime);
        const segs = timeDiff(dbak, dor) / 1000;
        return segs > 60;
    }

    function cleanCssInstall(): void
    {
        const c = fs
            .createReadStream(htmlFile)
            .pipe(fs.createWriteStream(htmlFileBack));
        c.on("finish", function ()
        {
            replaceCss();
        });
    }

    function installItem(bakfile: string, orfile: string, cleanInstallFunc: any): void
    {
        fs.stat(bakfile, function (errBak, statsBak)
        {
            //if (errBak)
            {
                // clean installation
                cleanInstallFunc();
              //  return;
            }

            // check htmlFileBack's timestamp and compare it to the htmlFile's.
            fs.stat(orfile, function (errOr, statsOr)
            {
                if (errOr)
                {
                    vscode.window.showInformationMessage(msg.smthingwrong + errOr);
                    return;
                }

                var updated = hasBeenUpdated(statsBak, statsOr);
                if (updated)
                {
                    // some update has occurred. clean install
                    cleanInstallFunc();
                }
            });
        });
    }

    function emitEndUninstall(): void
    {
        eventEmitter.emit("endUninstall");
    }

    function restoredAction(isRestored: number, willReinstall: boolean): void
    {
        if (isRestored >= 1)
        {
            if (willReinstall)
            {
                emitEndUninstall();
            } else
            {
                disabledRestart();
            }
        }
    }

    function restoreBak(willReinstall: boolean = false): void
    {
        var restore = 0;
        fs.unlink(htmlFile, function (err)
        {
            if (err)
            {
                vscode.window.showInformationMessage(msg.admin);
                return;
            }
            var c = fs
                .createReadStream(htmlFileBack)
                .pipe(fs.createWriteStream(htmlFile));
            c.on("finish", function ()
            {
                fs.unlinkSync(htmlFileBack);
                restore++;
                restoredAction(restore, willReinstall);
            });
        });
    }

    function reloadWindow(): void
    {
        // reload vscode-window
        vscode.commands.executeCommand("workbench.action.reloadWindow");
    }

    function enabledRestart(): void
    {
        vscode.window
            .showInformationMessage(msg.enabled, { title: msg.restartIde })
            .then(function (msg)
            {
                reloadWindow();
            });
    }

    function disabledRestart(): void
    {
        vscode.window
            .showInformationMessage(msg.disabled, { title: msg.restartIde })
            .then(function (msg)
            {
                reloadWindow();
            });
    }

    // ####  main commands ######################################################

    function fInstall(): void
    {
        installItem(htmlFileBack, htmlFile, cleanCssInstall);
    }

    function fUninstall(willReinstall: boolean = false): void
    {
        fs.stat(htmlFileBack, function (errBak, statsBak)
        {
            if (errBak)
            {
                if (willReinstall)
                {
                    emitEndUninstall();
                }
                return;
            }

            fs.stat(htmlFile, function (errOr, statsOr)
            {
                if (errOr)
                {
                    vscode.window.showInformationMessage(msg.smthingwrong + errOr);
                    return;
                }

                restoreBak(willReinstall);
            });
        });
    }

    function fUpdate(): void
    {
        eventEmitter.once("endUninstall", fInstall);
        fUninstall(true);
    }

    context.subscriptions.push(vscode.commands.registerCommand(
        "extension.installCustomDocumentWell", fInstall
    ));
    context.subscriptions.push(vscode.commands.registerCommand(
        "extension.uninstallCustomDocumentWell", fUninstall
    ));
    context.subscriptions.push(vscode.commands.registerCommand(
        "extension.updateCustomDocumentWell", fUpdate
    ));
}

// this method is called when your extension is deactivated
export function deactivate() { }
